buildFHSUserEnv: Allow having custom /opt in the FHS environment

wentasah opened the pull request:
###### Motivation for this change

buildFHSUserEnv is meant primarily for running 3rd-party software
which is difficult to patch for NixOS. Such software is often built to
run from /opt. Currently, running such a software from FHS environment
is difficult for two reasons:

1. If the host system has the /opt directory, it gets bind-mounted to
   the FHSEnv. If that directory is not accessible for non-root
   users (which is what docker's containerd does with /opt :-(), the
   user running the FHSEnv cannot use it.

   With the change in chrootenv.c, /opt is not bind-mounted to the
   container, but instead created as user-modifiable symlink to
   /host/opt (see the init attribute in
   build-fhs-userenv/default.nix). If needed, the user can remove this
   symlink and create an empty /opt directory which is under his/her
   control.

2. If the 3rd-party software is put into the Nix store via a simple
   derivation (with e.g. installPhase = "dpkg-deb -x $src $out"), the
   content of /opt directory of that derivation does not appear in the
   FHSEnv even if the derivation is specified in targetPkgs. This is
   why we change env.nix.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

wentasah committed:
buildFHSUserEnv: Allow having custom /opt in the FHS environment

buildFHSUserEnv is meant primarily for running 3rd-party software
which is difficult to patch for NixOS. Such software is often built to
run from /opt. Currently, running such a software from FHS environment
is difficult for two reasons:

1. If the 3rd-party software is put into the Nix store via a simple
   derivation (with e.g. installPhase = "dpkg-deb -x $src $out"), the
   content of /opt directory of that derivation does not appear in the
   FHSEnv even if the derivation is specified in targetPkgs. This is
   why we change env.nix.

2. If using buildFHSUserEnvChroot and the host system has the /opt
   directory, it always gets bind-mounted to the FHSEnv even if some
   targetPkgs contain /opt (NB buildFHSUserEnvBubblewrap does not have
   this problem). If that directory is not accessible for non-root
   users (which is what docker's containerd does with /opt :-(), the
   user running the FHSEnv cannot use it.

   With the change in chrootenv.c, /opt is not bind-mounted to the
   container, but instead created as user-modifiable symlink to
   /host/opt (see the init attribute in
   build-fhs-userenv/default.nix). If needed, the user can remove this
   symlink and create an empty /opt directory which is under his/her
   control.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 128126` at e5299631 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>ovftool</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/2a7b4c0207ece034603478cf5921e712">quartus-prime-lite</a></li>
  </ul>
</details>
<details>
  <summary>2 packages skipped due to time constraints:</summary>
  <ul>
    <li>houdini</li>
    <li>plex</li>
  </ul>
</details>
<details>
  <summary>28 packages built successfully:</summary>
  <ul>
    <li>appimage-run</li>
    <li>arduino-ci</li>
    <li>arduino-cli</li>
    <li>bitscope.chart</li>
    <li>bitscope.console</li>
    <li>bitscope.display</li>
    <li>bitscope.dso</li>
    <li>bitscope.logic</li>
    <li>bitscope.meter</li>
    <li>bitscope.proto</li>
    <li>bitscope.server</li>
    <li>clonehero</li>
    <li>conda</li>
    <li>dropbox</li>
    <li>dropbox-cli</li>
    <li>esphome</li>
    <li>fahclient</li>
    <li>flutter (flutterPackages.stable)</li>
    <li>hover</li>
    <li>left4gore-bin</li>
    <li>lightworks</li>
    <li>mate.caja-dropbox</li>
    <li>platformio</li>
    <li>shticker-book-unwritten</li>
    <li>sidequest</li>
    <li>unvanquished</li>
    <li>vmware-horizon-client</li>
    <li>zulip</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

jonringer commented:
Most new packages should be using the bubblewrap version, which I believe `buildFHSUserEnv` now points to. This is the `buildFHSUserChroot` helper.

wentasah commented:
`buildFHSUserEnv` still points to `buildFHSUserChroot`. Nevertheless, I updated `buildFHSUserEnvBubblewrap` in the same way and updated the commit message accordingly.

jonringer reviewed (approved):
LGTM

wentasah commented:
@SuperSandro2000 Is there anything more needed to get this merged?

Artturin commented:
Result of `nixpkgs-review pr 128126` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>minetime</li>
    <li>ovftool</li>
    <li>timedoctor</li>
  </ul>
</details>
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>appimage-run-tests</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>houdini</li>
    <li>quartus-prime-lite</li>
  </ul>
</details>
<details>
  <summary>88 packages built:</summary>
  <ul>
    <li>Sylk</li>
    <li>altair</li>
    <li>android-studio (androidStudioPackages.stable)</li>
    <li>androidStudioPackages.beta</li>
    <li>androidStudioPackages.canary</li>
    <li>androidStudioPackages.dev</li>
    <li>anki-bin</li>
    <li>appimage-run</li>
    <li>apple-music-electron</li>
    <li>arduino-ci</li>
    <li>arduino-cli</li>
    <li>betterdiscord-installer</li>
    <li>bitscope.chart</li>
    <li>bitscope.console</li>
    <li>bitscope.display</li>
    <li>bitscope.dso</li>
    <li>bitscope.logic</li>
    <li>bitscope.meter</li>
    <li>bitscope.proto</li>
    <li>bitscope.server</li>
    <li>bottles</li>
    <li>chrysalis</li>
    <li>clonehero</li>
    <li>conda</li>
    <li>crypto-org-wallet</li>
    <li>deltachat-electron</li>
    <li>devdocs-desktop</li>
    <li>dropbox</li>
    <li>dropbox-cli</li>
    <li>electron-mail</li>
    <li>electronplayer</li>
    <li>esphome</li>
    <li>fahclient</li>
    <li>flexoptix-app</li>
    <li>flutter (flutterPackages.stable)</li>
    <li>hover</li>
    <li>irccloud</li>
    <li>joplin-desktop</li>
    <li>keeweb</li>
    <li>kodiPackages.steam-launcher</li>
    <li>lbry</li>
    <li>ledger-live-desktop</li>
    <li>left4gore-bin</li>
    <li>lens</li>
    <li>lightworks</li>
    <li>losslesscut-bin</li>
    <li>lunar-client</li>
    <li>lutris</li>
    <li>lutris-free</li>
    <li>marktext</li>
    <li>mate.caja-dropbox</li>
    <li>mathpix-snipping-tool</li>
    <li>molotov</li>
    <li>mycrypto</li>
    <li>neo4j-desktop</li>
    <li>notable</li>
    <li>nuclear</li>
    <li>octant-desktop</li>
    <li>p3x-onenote</li>
    <li>platformio</li>
    <li>playonlinux</li>
    <li>plex</li>
    <li>plexamp</li>
    <li>protontricks</li>
    <li>radicle-upstream</li>
    <li>rambox-pro</li>
    <li>shticker-book-unwritten</li>
    <li>sidequest</li>
    <li>ssb-patchwork</li>
    <li>standardnotes</li>
    <li>station</li>
    <li>steam</li>
    <li>steam-run</li>
    <li>steam-run-native</li>
    <li>steam-tui</li>
    <li>steamcmd</li>
    <li>timeular</li>
    <li>trezor-suite</li>
    <li>tusk</li>
    <li>unityhub</li>
    <li>unvanquished</li>
    <li>vmware-horizon-client</li>
    <li>vscode-fhs</li>
    <li>vscodium-fhs</li>
    <li>wootility</li>
    <li>ytmdesktop</li>
    <li>zettlr</li>
    <li>zulip</li>
  </ul>
</details>
